consensus/core/src/dag_state.rs

```diff
@@ -426,20 +535,65 @@ impl DagState {
         }
     }
 
+    /// Detects and returns the blocks of the round that forms the last quorum. The method will return
+    /// the quorum even if that's genesis.
+    pub(crate) fn last_quorum(&self) -> Vec<VerifiedBlock> {
+        // the quorum should exist either on the highest accepted round or the one before. If we fail to detect
+        // a quorum then it means that our DAG has advanced with missing causal history.
+        for round in
+            (self.highest_accepted_round.saturating_sub(1)..=self.highest_accepted_round).rev()
+        {
+            if round == GENESIS_ROUND {
+                return self.genesis_blocks();
+            }
+            let mut quorum = StakeAggregator::<QuorumThreshold>::new();
+
+            // Since that the minimum wave length is 3 we expect to find a quorum in the uncommitted rounds.
+            let blocks = self.get_uncommitted_blocks_at_round(round);
+            for block in &blocks {
+                if quorum.add(block.author(), &self.context.committee) {
+                    return blocks;
+                }
+            }
+        }
+
+        panic!("Fatal error, no quorum has been detected in our DAG on the last two rounds.");
+    }
+
+    pub(crate) fn genesis_blocks(&self) -> Vec<VerifiedBlock> {
+        self.genesis.values().cloned().collect()
+    }
+
     /// Highest round where a block is committed, which is last commit's leader round.
     fn last_commit_round(&self) -> Round {
         match &self.last_commit {
             Some(commit) => commit.leader().round,
             None => 0,
         }
     }
+
+    /// The last round that got evicted after a cache clean up operation. After this round we are
+    /// guaranteed to have all the produced blocks from that authority. For any round that is
+    /// <= `last_evicted_round` we don't have such guarantees as out of order blocks might exist.
+    fn authority_evict_round(&self, authority_index: AuthorityIndex) -> Round {
+        let commit_round = self.last_committed_rounds[authority_index];
+        Self::evict_round(commit_round, self.cached_rounds)
+    }
+
+    /// Calculates the last eviction round based on the provided `commit_round`. Any blocks with
+    /// round <= the evict round have been cleaned up.
+    fn evict_round(commit_round: Round, cached_rounds: Round) -> Round {
+        commit_round.saturating_sub(cached_rounds).saturating_sub(1)
```

Comment:
I don't think the last `saturating_sub(1)` is necessary? Suppose last committed rounds is 300, and "cached rounds" are 100. Evicting rounds <= 200 is correct. Rounds from 201~300 are cached which are indeed 100 rounds.